File: csharp/HearthstoneReplays/Parser/Regexes.cs
#region

using System.Text.RegularExpressions;

#endregion

namespace HearthstoneReplays.Parser
{
	public class Regexes
	{
		public static readonly Regex PowerlogLineRegex = new Regex(@"^D ([\d:.]+) ([^(]+)\(\) - (.+)$");
		public static readonly Regex OutputlogLineRegex = new Regex(@"\[Power\] ()([^(]+)\(\) - (.+)$");
		public static readonly Regex EntityRegex = new Regex(@"\[.*\s*id=(\d+)\s*.*\]");

		public static readonly Regex ChoicesChoiceRegex =
			new Regex(@"id=(\d+) PlayerId=(\d+) ChoiceType=(\w+) CountMin=(\d+) CountMax=(\d+)$");

		public static readonly Regex ChoicesSourceRegex = new Regex(@"Source=(\[?.+\]?)$");
		public static readonly Regex ChoicesEntitiesRegex = new Regex(@"Entities\[(\d+)\]=(\[.+\])$");
		public static readonly Regex SendChoicesChoicetypeRegex = new Regex(@"id=(\d+) ChoiceType=(.+)$");
		public static readonly Regex SendChoicesEntitiesRegex = new Regex(@"m_chosenEntities\[(\d+)\]=(\[.+\])$");
		public static readonly Regex OptionsEntityRegex = new Regex(@"id=(\d+)$");
		public static readonly Regex OptionsOptionRegex = new Regex(@"option (\d+) type=(\w+) mainEntity=(.*)$");
		public static readonly Regex OptionsSuboptionRegex = new Regex(@"(subOption|target) (\d+) entity=(.*)$");

		public static readonly Regex SendOptionRegex =
			new Regex(@"selectedOption=(\d+) selectedSubOption=(-1|\d+) selectedTarget=(\d+) selectedPosition=(\d+)");

		public static readonly Regex ActionTagRegex = new Regex(@"tag=(\w+) value=(\w+)");
		public static readonly Regex ActionFullentityRegex1 = new Regex(@"FULL_ENTITY - Updating (\[.+\]) CardID=(\w+)?$");
		public static readonly Regex ActionFullentityRegex2 = new Regex(@"FULL_ENTITY - Creating ID=(\d+) CardID=(\w+)?$");
		public static readonly Regex ActionShowEntityRegex = new Regex(@"SHOW_ENTITY - Updating Entity=(\[?.+\]?) CardID=(\w+)$");
		public static readonly Regex ActionHideEntityRegex = new Regex(@"HIDE_ENTITY - Entity=(\[?.+\]?) tag=(\w+) value=(\w+)");
		public static readonly Regex ActionTagChangeRegex = new Regex(@"TAG_CHANGE Entity=(\[?.+\]?) tag=(\w+) value=(\w+)");

		public static readonly Regex ActionStartRegex =
			new Regex(@"ACTION_START Entity=(\[?.+\]?) (?:SubType|BlockType)=(\w+) Index=(-1|\d+) Target=(\[?.+\]?)$");

		public static readonly Regex ActionMetadataRegex = new Regex(@"META_DATA - Meta=(\w+) Data=(\[?.+\]?) Info=(\d+)");
		public static readonly Regex ActionCreategameRegex = new Regex(@"GameEntity EntityID=(\d+)");

		public static readonly Regex ActionCreategamePlayerRegex =
			new Regex(@"Player EntityID=(\d+) PlayerID=(\d+) GameAccountId=\[hi=(\d+) lo=(\d+)\]$");

		public static readonly Regex ActionMetaDataInfoRegex = new Regex(@"Info\[(\d+)\] = (\[?.+\]?)");
	}
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/Regexes.java
package info.hearthsim.hsreplay.parser;

import java.text.MessageFormat;
import java.util.regex.Pattern;

public class Regexes {

	private static final String Entity = "(GameEntity|UNKNOWN HUMAN PLAYER|\\[.+\\]|\\d+|.+)";

	public static final Pattern PowerlogLineRegex = Pattern.compile("^D ([\\d:.]+) ([^(]+)\\(\\) - (.+)$");
	public static final Pattern OutputlogLineRegex = Pattern.compile("\\[Power\\] ()([^(]+)\\(\\) - (.+)$");
	public static final Pattern EntityRegex = Pattern.compile("\\[.*\\s*id=(\\d+)\\s*.*\\]");

	public static final Pattern ChoicesChoiceRegex_OLD = Pattern
			.compile("id=(\\d+) PlayerId=(\\d+) ChoiceType=(\\w+) CountMin=(\\d+) CountMax=(\\d+)$");
	public static final Pattern ChoicesChoiceRegex = Pattern.compile(MessageFormat.format(
			"id=(\\d+) Player={0} TaskList=(\\d+)? ChoiceType=(\\w+) CountMin=(\\d+) CountMax=(\\d+)$", Entity));
	public static final Pattern ChoicesSourceRegex = Pattern.compile(MessageFormat.format("Source={0}$", Entity));
	public static final Pattern ChoicesEntitiesRegex = Pattern.compile("Entities\\[(\\d+)\\]=(\\[.+\\])$");

	public static final Pattern ActionCreategameRegex = Pattern.compile("GameEntity EntityID=(\\d+)");
	public static final Pattern ActionCreategamePlayerRegex = Pattern
			.compile("Player EntityID=(\\d+) PlayerID=(\\d+) GameAccountId=\\[hi=(\\d+) lo=(\\d+)\\]$");
	public static final Pattern ActionStartRegex = Pattern.compile(MessageFormat
			.format("ACTION_START Entity={0} (?:SubType|BlockType)=(\\w+) Index=(-1|\\d+) Target={0}$", Entity));
	public static final Pattern ActionMetadataRegex = Pattern
			.compile(MessageFormat.format("META_DATA - Meta=(\\w+) Data={0} Info=(\\d+)", Entity));
	public static final Pattern ActionMetaDataInfoRegex = Pattern
			.compile(MessageFormat.format("Info\\[(\\d+)\\] = {0}", Entity));
	public static final Pattern ActionShowEntityRegex = Pattern
			.compile(MessageFormat.format("SHOW_ENTITY - Updating Entity={0} CardID=(\\w+)$", Entity));
	public static final Pattern ActionHideEntityRegex = Pattern
			.compile(MessageFormat.format("HIDE_ENTITY - Entity={0} tag=(\\w+) value=(\\w+)", Entity));
	public static final Pattern ActionFullEntityUpdatingRegex = Pattern
			.compile(MessageFormat.format("FULL_ENTITY - Updating {0} CardID=(\\w+)?$", Entity));
	public static final Pattern ActionFullEntityCreatingRegex = Pattern
			.compile("FULL_ENTITY - Creating ID=(\\d+) CardID=(\\w+)?$");
	public static final Pattern ActionTagChangeRegex = Pattern
			.compile(MessageFormat.format("TAG_CHANGE Entity={0} tag=(\\w+) value=(\\w+)", Entity));
	public static final Pattern ActionTagRegex = Pattern.compile("tag=(\\w+) value=(\\w+)");

	public static final Pattern EntitiesChosenRegex = Pattern
			.compile(MessageFormat.format("id=(\\d+) Player={0} EntitiesCount=(\\d+)$", Entity));
	public static final Pattern EntitiesChosenEntitiesRegex = Pattern
			.compile(MessageFormat.format("Entities\\[(\\d+)\\]={0}$", Entity));

	public static final Pattern OptionsEntityRegex = Pattern.compile("id=(\\d+)$");
	public static final Pattern OptionsOptionRegex = Pattern
			.compile(MessageFormat.format("option (\\d+) type=(\\w+) mainEntity={0}?$", Entity));
	public static final Pattern OptionsSuboptionRegex = Pattern
			.compile(MessageFormat.format("(subOption|target) (\\d+) entity={0}?$", Entity));

	public static final Pattern SendChoicesChoicetypeRegex = Pattern.compile("id=(\\d+) ChoiceType=(.+)$");
	public static final Pattern SendChoicesEntitiesRegex = Pattern.compile("m_chosenEntities\\[(\\d+)\\]=(\\[.+\\])$");

	public static final Pattern SendOptionRegex = Pattern
			.compile("selectedOption=(\\d+) selectedSubOption=(-1|\\d+) selectedTarget=(\\d+) selectedPosition=(\\d+)");

}
